File: day-2-pm/sandbox/tests/lexer_test.cpp
#include "lexer.h"

#include <gtest/gtest.h>

using calc::Lexer;
using calc::TokenType;

TEST(LexerTest, SingleNumber) {
    Lexer lexer("42");
    auto tok = lexer.next();
    EXPECT_EQ(tok.type, TokenType::NUMBER);
    EXPECT_EQ(tok.value, 42);
    EXPECT_EQ(lexer.next().type, TokenType::END);
}

TEST(LexerTest, AllOperators) {
    Lexer lexer("+-*/");
    EXPECT_EQ(lexer.next().type, TokenType::PLUS);
    EXPECT_EQ(lexer.next().type, TokenType::MINUS);
    EXPECT_EQ(lexer.next().type, TokenType::STAR);
    EXPECT_EQ(lexer.next().type, TokenType::SLASH);
    EXPECT_EQ(lexer.next().type, TokenType::END);
}

TEST(LexerTest, SkipsWhitespaceAndMixedExpr) {
    Lexer lexer("  12 + 30 ");
    auto a = lexer.next();
    EXPECT_EQ(a.type, TokenType::NUMBER);
    EXPECT_EQ(a.value, 12);
    EXPECT_EQ(lexer.next().type, TokenType::PLUS);
    auto b = lexer.next();
    EXPECT_EQ(b.type, TokenType::NUMBER);
    EXPECT_EQ(b.value, 30);
    EXPECT_EQ(lexer.next().type, TokenType::END);
}

TEST(LexerTest, EndIsIdempotent) {
    Lexer lexer("");
    EXPECT_EQ(lexer.next().type, TokenType::END);
    EXPECT_EQ(lexer.next().type, TokenType::END);
}

TEST(LexerTest, UnknownCharacterThrows) {
    Lexer lexer("1 ? 2");
    EXPECT_EQ(lexer.next().type, TokenType::NUMBER);
    EXPECT_THROW(lexer.next(), std::runtime_error);
}

File: day-2-pm/sandbox-java/src/test/java/calc/LexerTest.java
package calc;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.Test;

class LexerTest {

    @Test
    void SingleNumber() {
        Lexer lexer = new Lexer("42");
        Token tok = lexer.next();
        assertEquals(TokenType.NUMBER, tok.type);
        assertEquals(42, tok.value);
        assertEquals(TokenType.END, lexer.next().type);
    }

    @Test
    void AllOperators() {
        Lexer lexer = new Lexer("+-*/");
        assertEquals(TokenType.PLUS, lexer.next().type);
        assertEquals(TokenType.MINUS, lexer.next().type);
        assertEquals(TokenType.STAR, lexer.next().type);
        assertEquals(TokenType.SLASH, lexer.next().type);
        assertEquals(TokenType.END, lexer.next().type);
    }

    @Test
    void SkipsWhitespaceAndMixedExpr() {
        Lexer lexer = new Lexer("  12 + 30 ");
        Token a = lexer.next();
        assertEquals(TokenType.NUMBER, a.type);
        assertEquals(12, a.value);
        assertEquals(TokenType.PLUS, lexer.next().type);
        Token b = lexer.next();
        assertEquals(TokenType.NUMBER, b.type);
        assertEquals(30, b.value);
        assertEquals(TokenType.END, lexer.next().type);
    }

    @Test
    void EndIsIdempotent() {
        Lexer lexer = new Lexer("");
        assertEquals(TokenType.END, lexer.next().type);
        assertEquals(TokenType.END, lexer.next().type);
    }

    @Test
    void UnknownCharacterThrows() {
        Lexer lexer = new Lexer("1 ? 2");
        assertEquals(TokenType.NUMBER, lexer.next().type);
        assertThrows(CalcException.class, () -> lexer.next());
    }
}
